controller/aks-cluster-config-handler.go

```diff
@@ -707,21 +709,36 @@ func (h *Handler) updateUpstreamClusterState(ctx context.Context, secretsCache w
 		return config, err
 	}
 
-	// check tags for update
-	if config.Spec.Tags != nil {
+	// check tags for update. Skip the tags update if there was recently an unexpected response from Azure for a tags
+	// update so the cluster can become active.
+	if config.Spec.Tags != nil && !UnexpectedTagUpdateResponseInAzure {
+
 		if !reflect.DeepEqual(config.Spec.Tags, upstreamSpec.Tags) {
 			logrus.Infof("Updating tags for cluster [%s]", config.Spec.ClusterName)
 			tags := containerservice.TagsObject{
 				Tags: *to.StringMapPtr(config.Spec.Tags),
 			}
-			_, err = resourceClusterClient.UpdateTags(ctx, config.Spec.ResourceGroup, config.Spec.ClusterName, tags)
+			response, err := resourceClusterClient.UpdateTags(ctx, config.Spec.ResourceGroup, config.Spec.ClusterName, tags)
 			if err != nil {
 				return config, err
 			}
+			// Azure will regenerate default upstream tags and add them to the cluster resource if a user tries to
+			// remove them all in one request, but the Azure job that removes tags doesn't return an error in that
+			// case. So, we check whether Azure returns an unexpected state for the tags and if so, log the response
+			// and move on. Any upstream tags that were regenerated and put back on the cluster by Azure will be synced
+			// back to rancher.
+			if !reflect.DeepEqual(tags, upstreamSpec.Tags) && response.Status() == "Succeeded" {
+				logrus.Infof("Azure API response was not expected. Completing tag update for cluster [%s]", config.Spec.ClusterName)
+				UnexpectedTagUpdateResponseInAzure = true
+			}
 			return h.enqueueUpdate(config)
 		}
 	}
 
+	if UnexpectedTagUpdateResponseInAzure {
+		UnexpectedTagUpdateResponseInAzure = false
+	}
```

Comment:
If a user has more than one AKS cluster, then this global variable gets used for all of them. That means that if one cluster has an issue updating tags, then another cluster may have its tag update inadvertently skipped. Which would cause issues for the cluster having trouble updating its tags.

Reply:
Good thought! Testing..